### PR TITLE
Hack to work around a race in Smack.

### PR DESCRIPTION
**Pull request by JonathanLennox**
(See commit 3f570999 for an explanation of the race.)

### CODE REVIEW

**Review by bgrozev — DISMISSED** [review]

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java [hunk 1]
```diff
@@ -1076,4 +1081,28 @@ public void roomDestroyed(MultiUserChat alternateMUC, String reason)
...
+        assert(!muc.isJoined());
+
+        Field occupantsMapField = null;
+        try
+        {
+            occupantsMapField = muc.getClass().getField("occupantsMap");
+            occupantsMapField.setAccessible(true);
+
+            Map<EntityFullJid, Presence> occupantsMap = (Map<EntityFullJid, Presence>)occupantsMapField.get(muc);
+            occupantsMap.clear();
```
Comment (damencho) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java:1101]: What about printing a log if there was something in the map?

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java [hunk 2]
```diff
@@ -1076,4 +1081,28 @@ public void roomDestroyed(MultiUserChat alternateMUC, String reason)
...
+     * chat instance.  This function uses reflection to hack the internal state to fix the problem.
+     */
+    private void clearMucOccupantsMap(MultiUserChat muc)
+    {
+        assert(!muc.isJoined());
+
+        Field occupantsMapField = null;
+        try
+        {
+            occupantsMapField = muc.getClass().getField("occupantsMap");
```
Comment (damencho) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java:1097]: This should be getDeclaredField

**Comment by damencho** [comment]
I did some automation and reverted the previous change and not I cannot hit it ... it is so strange ... I even add in prosody code to send a self presence update before the unavailable message and I see it in jicofo xmpp logs but still nothing, I ran 50 attempts ... will leave it now to do more attempts during the day. 
I'm waiting for a print I added on `if (muc.getOccupants().size() > 0)` or at least the test will fail to join the muc and I will see that ...

**Comment by damencho** [comment]
Ok, here is an update. Out of 200 runs managed to repro it once and the fix with clearing the map worked, no locking. 
The runs were without prosody modifications and just with reverted the latest change in master about the setting of bridge 0.

**Comment by damencho** [comment]
And leaving just the fix from Friday I was able to repro in around 60 attempts. So I think we need this one :)

**Review by bgrozev — APPROVED** [review]

**Review by damencho — APPROVED** [review]

**Pull request merged**